File: TUnit.TestProject/NotInParallelTests.cs
using System.Collections.Concurrent;
using TUnit.TestProject.Attributes;

namespace TUnit.TestProject;

[EngineTest(ExpectedResult.Pass)]
[Retry(3)]
public class NotInParallelTests
{
    private static readonly ConcurrentBag<DateTimeRange> TestDateTimeRanges = [];

    [After(Test)]
    public async Task TestOverlaps()
    {
        TestDateTimeRanges.Add(new DateTimeRange(TestContext.Current!.Execution.TestStart!.Value, TestContext.Current.Execution.Result!.End!.Value));

        await AssertNoOverlaps();
    }

    [Test, NotInParallel, Repeat(3)]
    public async Task NotInParallel_Test1(CancellationToken cancellationToken)
    {
        var timeProvider = TimeProviderContext.Current;
        await timeProvider.Delay(TimeSpan.FromMilliseconds(500), cancellationToken);
    }

    [Test, NotInParallel, Repeat(3)]
    public async Task NotInParallel_Test2(CancellationToken cancellationToken)
    {
        var timeProvider = TimeProviderContext.Current;
        await timeProvider.Delay(TimeSpan.FromMilliseconds(500), cancellationToken);
    }

    [Test, NotInParallel, Repeat(3)]
    public async Task NotInParallel_Test3(CancellationToken cancellationToken)
    {
        var timeProvider = TimeProviderContext.Current;
        await timeProvider.Delay(TimeSpan.FromMilliseconds(500), cancellationToken);
    }

    private async Task AssertNoOverlaps()
    {
        foreach (var testDateTimeRange in TestDateTimeRanges)
        {
            await Assert.That(TestDateTimeRanges
                .Except([testDateTimeRange])
                .Any(x => x.Overlap(testDateTimeRange)))
                .IsFalse();
        }
    }

    private class DateTimeRange(DateTimeOffset start, DateTimeOffset end)
    {
        public DateTimeOffset Start { get; } = start;
        public DateTimeOffset End { get; } = end;

        public bool Overlap(DateTimeRange other)
        {
            return Start <= other.End && other.Start <= End;
        }
    }
}

File: TUnit.TestProject.FSharp/NotInParallelTests.fs
namespace TUnit.TestProject.FSharp

open System
open System.Collections.Concurrent
open System.Linq
open System.Threading.Tasks
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Core

// F# equivalent of NotInParallelTests.cs

type DateTimeRange(start: DateTime, end_: DateTime) =
    member _.Start = start
    member _.End = end_
    member this.Overlap(other: DateTimeRange) =
        this.Start <= other.End && other.Start <= this.End

type NotInParallelTests() =
    static let testDateTimeRanges = ConcurrentBag<DateTimeRange>()

    [<After(HookType.Test)>]
    member _.TestOverlaps() : Task = task {
        testDateTimeRanges.Add(DateTimeRange(TestContext.Current.TestStart.Value.DateTime, TestContext.Current.Result.End.Value.DateTime))
        do! NotInParallelTests.AssertNoOverlaps()
    }

    [<Test>]
    [<NotInParallel>]
    [<Repeat(3)>]
    member _.NotInParallel_Test1() : Task =
        Task.Delay(500)

    [<Test>]
    [<NotInParallel>]
    [<Repeat(3)>]
    member _.NotInParallel_Test2() : Task =
        Task.Delay(500)

    [<Test>]
    [<NotInParallel>]
    [<Repeat(3)>]
    member _.NotInParallel_Test3() : Task =
        Task.Delay(500)

    static member private AssertNoOverlaps() : Task = task {
        for testDateTimeRange in testDateTimeRanges do
            let overlaps = testDateTimeRanges
                           |> Seq.filter (fun x -> not (Object.ReferenceEquals(x, testDateTimeRange)))
                           |> Seq.exists (fun x -> x.Overlap(testDateTimeRange))
            do! Assert.That(overlaps).IsFalse()
    }
